### PR TITLE
feat: allow multiple kepler deployments using kepler-internal

### PR DESCRIPTION
**Pull request by sthaha**
**Important Changes:**
* allows setting `deployment-namespace` instead of hardcoded `openshift-kepler-operator`
* modifies the default kepler deployment namespace to `kepler-operator` 
* Allows watching additional namespaces using `watch-namepaces=foo,bar --watch-namespaces=baz` (multiple list)
* Adds `status.exporter` to indicate kepler deployment `conditions` 
* 

** Testing the PR locally** 
To give this a spin on `kind` cluster created using `make fresh` 

```sh
operator-sdk run bundle quay.io/sthaha/kepler-operator-bundle:0.0.0-kepler-internal \
  --namespace=operators \
  --use-http
```

### CODE REVIEW

File: pkg/components/exporter/exporter.go [hunk 1]
```diff
@@ -37,40 +36,29 @@ import (
...
+	// FQServiceAccountName = "system:serviceaccount:" + k.Namespace() + ":" + ServiceAccountName
+	//
+	// ClusterRoleName        = prefix + "cr"
+	// ClusterRoleBindingName = prefix + "crb"
 
-	ClusterRoleName        = prefix + "cr"
-	ClusterRoleBindingName = prefix + "crb"
-
-	ConfigmapName = prefix + "cm"
-	DaemonSetName = prefix + "ds"
```
Comment (vprashar2929) [inline pkg/components/exporter/exporter.go:51]: Shouldn't we keep `ds` name constant as `kepler-exporter` ?

**Comment by vprashar2929** [comment]
In case of multiple kepler deployments we have to figure out  openshift dashboard's deployments and also revise our prometheus rules accordingly.

File: pkg/components/exporter/exporter.go (hunk 1 above)
Comment (sthaha) [reply pkg/components/exporter/exporter.go:51]: what would happen if the 2 keplerinternals are deployed to the the same namespace with the daemonset name set as a const 💭 ?

**Comment by sthaha** [comment]
> In case of multiple kepler deployments we have to figure out openshift dashboard's deployments and also revise our prometheus rules accordingly.

Thats not a supported scenario to start with. I am thinking of adding a
```
spec:
  dashboards: 
      enabled: t|f
  exporter:
     ....
```
to KeplerInternal where `dashboards.enabled` defaults to `false`. I also think the prometheusrule may be giving wrong values since we are not filter out by service.

**Comment by sthaha** [comment]
@vprashar2929 I have added an `openshift` specific api now to enable / disable dashboards.

File: pkg/api/v1alpha1/kepler_internal_types.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+type DashboardSpec struct {
+	// +kubebuilder:default=false
+	Enabled bool `json:"enabled,omitempty"`
+}
+
+type OpenShiftSpec struct {
+	// +kubebuilder:default=false
+	Enabled   bool          `json:"enabled,omitempty"`
+	Dashboard DashboardSpec `json:"dashboard,omitempty"`
```
Comment (vimalk78) [inline pkg/api/v1alpha1/kepler_internal_types.go:51]: for openshift, can we keep the `Enabled` default value to true?

File: cmd/manager/main.go [hunk 3]
```diff
@@ -59,26 +59,45 @@ func init() {
...
 	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8080", "The address the metric endpoint binds to.")
 	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
 	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
 
+	flag.StringVar(&controllers.KeplerDeploymentNS, "deployment-namespace", controllers.KeplerDeploymentNS,
+		"Namespace where kepler and its components are deployed.")
+
+	flag.CommandLine.Var(flag.Value(&additionalNamespaces), "watch-namespaces",
```
Comment (vimalk78) [inline cmd/manager/main.go:90]: can we elaborate a bit how this parameter is to be used? is this for running multiple kepler-exporters?

File: bundle/manifests/kepler-operator.clusterserviceversion.yaml [hunk 4]
```diff
@@ -340,5 +358,5 @@ spec:
     name: kepler
   - image: quay.io/sustainable_computing_io/kepler:release-0.6.1-libbpf
     name: kepler-libbpf
-  replaces: kepler-operator.v0.9.0
+  replaces: kepler-operator.v0.0.0-kepler-internal
```
[the remark this answers is not in the document]
Comment (sthaha) [reply bundle/manifests/kepler-operator.clusterserviceversion.yaml:361]: Thank you! good catch

File: cmd/manager/main.go (hunk 3 above)
Comment (sthaha) [reply cmd/manager/main.go:90]: I am also thinking of allowing a `*` as one of the options to allow watching all namespaces but lets do it when we really find a need.

File: pkg/api/v1alpha1/kepler_internal_types.go (hunk 2 above)
Comment (sthaha) [reply pkg/api/v1alpha1/kepler_internal_types.go:51]: Sure ... Internal API can break at any time :D

File: pkg/components/exporter/exporter.go [hunk 5]
```diff
@@ -94,57 +57,62 @@ var (
...
 			},
 		}
 	}
 
-	deployment := k.Spec.Exporter.Deployment
+	deployment := k.Spec.Exporter.Deployment.ExporterDeploymentSpec
+	image := k.Spec.Exporter.Deployment.Image
 	nodeSelector := deployment.NodeSelector
 	tolerations := deployment.Tolerations
+	port := deployment.Port
```
Comment (vimalk78) [inline pkg/components/exporter/exporter.go:80]: for multiple kepler deployments, the `port` also has to be different. Users must ensure that.

File: bundle/manifests/kepler-operator.clusterserviceversion.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (vimalk78) [reply bundle/manifests/kepler-operator.clusterserviceversion.yaml:361]: it was easiest to catch  :D

File: tests/e2e/kepler_test.go
Comment (vimalk78) [inline tests/e2e/kepler_test.go:1]: can we add a similar e2e test for keplerinternal ?

Comment (vimalk78) [reply tests/e2e/kepler_test.go:1]: may be in another PR.

**Review by vimalk78 — APPROVED** [review]
lgtm

**Pull request merged**